File: src/linkedList/python/palindrome/frequencyApproach.py
import collections

from linkedList.python.node import node


def is_palindrome(curr: node) -> bool:
    contents = collections.defaultdict(int)
    while curr:
        contents[curr.data] += 1
        curr = curr.next

    one_diff = False
    for v in contents.values():
        if v % 2 == 1:
            if one_diff:
                return False
            one_diff = True
    return True

File: src/linkedList/java/palindrome/frequencyApproach.java
package linkedList.java.palindrome;

import linkedList.java.Node;

import java.util.Map;
import java.util.HashMap;

public class frequencyApproach {
	public boolean isPalindrome(Node curr){
		Map<Integer, Integer> hm = new HashMap<>();
		while(curr != null){
			hm.computeIfPresent(curr.data, (k, v) -> v + 1);
			hm.putIfAbsent(curr.data, 1);
			curr = curr.next;
		}

		boolean oneDiff = false;
		for(Integer i: hm.values()){
			if(i % 2 == 1){
				if(oneDiff) return false;
				oneDiff = true;
			}
		}
		return true;
	}
}
